### DISCUSSION TITLE
Customizing TSML via Wordpress Action Hook

### DISCUSSION THREAD

**Discussion by brotskydotcom** [code4recovery/12-step-meeting-list#1590] [Show and tell]
### TL;DR
If you are running your TSML customizations off a Wordpress action hook, you need to use the `init` hook to get them to take effect properly. Until a recent release, you could have used the `after_setup_theme` hook, but no more.

### Long version:

I recently debugged and resolved a problem customizing TSML, and wanted to share the resolution in case others have the same issue.

In the [TSML FAQ](https://wordpress.org/plugins/12-step-meeting-list/#faq-header), a number of references are made to customizing TSML by adding snippets to your `functions.php` file. I don't like to put code in that file, because it is often updated during theme and other site updates, so custom code often gets lost. Instead, I inject my custom code using the free version of the [Code Snippets plugin](https://wordpress.org/plugins/code-snippets/).

Code injected using Snippets is executed when the snippet plugin loads, which is generally not when you want your customization code to execute. To get it to execute at the right time, you write your snippet so that it runs your code from one of the [Wordpress Action Hooks](https://developer.wordpress.org/plugins/hooks/actions/), like this:
```php
add_hook('after_setup_theme', function () {
    // do your customization here
});
```
In fact, this is exactly what my snippet looked like, and it was working great until a few releases ago, when it suddenly stopped working. My overrides of things like meeting types and their descriptions were no longer showing up on [my site](https://al-anonbythebay.org). Instead, I was getting the default meeting types and descriptions.

After some debugging, I figured out that the hook I was using (`after_setup_theme`), which had been running *after* TSML initialization, was now running *before* TSML initialization. Thus, my changes were being overwritten by the TSML default values. The fix was to run on the `init` hook instead of the `after_setup_theme` hook. With that one change, everything is working perfectly again. That is, instead of the code above, my code now looks like this:
```php
add_hook('init', function () {
    // do your customization here
});
```

**Comment by joshreisner** [comment]
thanks for the insight! another option i've found that works in some cases is to add customizations to `wp-config.php` since that file is never overwritten

**Comment by ginslo** [comment]
Another option to keep your customizations safe during updates is to create a child theme. The child theme inherits the look and functionality of its parent but does not get changed during theme updates. You would add your changes to the child's `functions.php` file rather than the parent's. Don't know for sure if that would solve the sequence issue, where your hook runs before TSML Initializaton or not but I believe the sequence would be the same as if it was in the parent theme.

**Comment by k8fish3r** [comment]
Could i ask the type of adjustments you've achieved this way? I used to adjust the meeting details pages (old version TSML) but haven't found a route to reapply these adjustments.

Are you applying adjustments to the main list, or meeting detail view?

Thank you for sharing.

**Reply by brotskydotcom** [reply]
I'm just adjusting the `tsml_custom_types`, `tsml_custom_flags`, and `tsml_custom_descriptions` in my snippet, following the instructions for doing this in the [FAQ](https://wordpress.org/plugins/12-step-meeting-list/#faq-header).

These changes affect both the meeting list and the individual listings.

My only other customization is not a code snippet, but the CSS snippet described [here](https://github.com/code4recovery/tsml-ui/issues/109). For that I just use the standard Wordpress custom CSS controls.

**Comment by brotskydotcom** [comment]
Update: as of version 3.19 of the TSML plugin, running in the `init` hook is _too late_ to update the customization of the TSML plugin. I now have success running my customization action on the `plugins_loaded` hook (which is where I would have expected it to work in the first place).

@joshreisner I would like to suggest that the TSML plugin have its own custom hook that can be used to apply these customizations. That would allow people to know, no matter where they end up adding their code, that it will be run at exactly the right time.
